http_server/request_testing/request_test.go

```diff
@@ -0,0 +1,112 @@
+package requesttesting
+
+import (
+	"context"
+	"log"
+	"net"
+	"net/http"
+	"testing"
+)
+
+type handlerRecorder struct {
+	req         *http.Request
+	finishedReq chan bool
+	//TODO(mara): add server response
+}
+
+
+func (hr *handlerRecorder) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	hr.req = r
+	log.Print("Handler was called.")
+	// log.Print(hr.req.URL.Query()["vegetable"])
+	hr.finishedReq <- true
+}
+
+func makeRequest(req string) *http.Request {
+	e1, e2 := net.Pipe()
+	listener := newMockListener(e2)
+	done := make(chan bool)
+	defer listener.Close()
+	handler := &handlerRecorder{req: nil, finishedReq: done}
+	server := &http.Server{Handler: handler}
+	go server.Serve(&listener)
+	defer server.Shutdown(context.Background())
+	e1.Write([]byte(req))
+	<-handler.finishedReq
+	return handler.req
+}
+
+func TestSetup(t *testing.T) {
+	req := makeRequest("GET / HTTP/1.1\r\n" +
+		"Host: localhost:8080\r\n" + "\r\n")
+
+	if req == nil {
+		t.Error("Expected GET Request, received nil")
+	}
```

Comment:
This cannot happen. If a request is not received you never send over the channel, so the call to `makeRequest` never returns. In order for this to manifest you'd need to add a timeout to your makeRequest call.

Reply:
is it a good idea to use timeout?